### PR TITLE
Add implicit animations codelab

### PR DESCRIPTION
**Pull request by legalcodes**
Staged: 
https://jt-flutter-website-2.firebaseapp.com/docs/codelabs/implicit-animations

Jump to markdown file:
https://github.com/flutter/website/pull/3118/files#diff-0b95e489513499b71bf4745f5b7bea31

### CODE REVIEW

File: src/docs/codelabs/implicit-animations.md [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+
+### Animating multiple properties
+
+In the preceding example, each property in the `Container` widget (`color`, 
+`borderRadius`, and `margin`) has an associated function (`randomColor`,
+`randomBorderRadius`, and `randomMargin` respectively) that returns a randomly
+generated value for that property. By using an `AniamtedContainer` widget, you can 
+quickly refactor this code to:
+
+- Generate new values for `color`, `borderRadius`, and `margin` whenever the user clicks the "change button".
```
Comment (sfshaza2) [inline src/docs/codelabs/implicit-animations.md]: "change button" => **Change** button

File: src/docs/codelabs/implicit-animations.md [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+  frameborder="no" height="800" width="100%"
+></iframe>
+
+### Animating multiple properties
+
+In the preceding example, each property in the `Container` widget (`color`, 
+`borderRadius`, and `margin`) has an associated function (`randomColor`,
+`randomBorderRadius`, and `randomMargin` respectively) that returns a randomly
+generated value for that property. By using an `AniamtedContainer` widget, you can 
+quickly refactor this code to:
```
Comment (sfshaza2) [inline src/docs/codelabs/implicit-animations.md]: Sentence fragment. => "refactor this code to do the following:"

File: src/docs/codelabs/implicit-animations.md [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+                 style: TextStyle(color: Colors.white),
+               ),
+-              onPressed: () => null,
++              onPressed: () => change(),
+             ),
+           ],
+         ),
+```
+
+**5.** Finally, set the duration of the animation that will power the transition
```
Comment (sfshaza2) [inline src/docs/codelabs/implicit-animations.md]: "that will power"  => "that powers"

File: src/docs/codelabs/implicit-animations.md [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+                   borderRadius: BorderRadius.circular(borderRadius),
+                 ),
+               ),
++              duration: _duration,
+             ),
+             MaterialButton(
+               color: Theme.of(context).primaryColor,
+```
+
+Here’s the example with the completed changes you’ve just made – run the code
```
Comment (sfshaza2) [inline src/docs/codelabs/implicit-animations.md]: &mdash;

File: src/docs/codelabs/implicit-animations.md [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+  src="{{site.custom.dartpadx.embed-flutter-prefix}}?id=ddfbc68ec9dc28a48703d29248f5366f"
+  style="border: 1px solid lightgrey; margin-top: 10px; margin-bottom: 25px" 
+  frameborder="no" height="800" width="100%"
+></iframe>
+
+### Changing an animation curve
+
+The preceding example demonstrates animated transitions between values for 
+the container's `margin`, `borderRadius`, and `color` properties. However, 
+`AnimatedContainer` animates changes in all of its properties, including ones
```
Comment (sfshaza2) [inline src/docs/codelabs/implicit-animations.md]: Not a fan of "ones":
"including ones..."  => including those that...

File: src/docs/codelabs/implicit-animations.md [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+we didn't use such as `padding`, `transform`, and even `child` and `alignment`!
+
+Additionally, all implicit animations allow you to pass a [curve] parameter
+used to control how an animation changes its value over the course of the
+specified duration. In the preceding examples so far, you did not specify a 
+curve for any of the implicit animations, so the implicit animations applied a 
+[linear animation curve] by default.
+
+However, watch how the animation changes in the preceding example
+when you pass a `fastOutSlowIn` curve:
```
Comment (sfshaza2) [inline src/docs/codelabs/implicit-animations.md]: Hmmmm... I'm of two minds about this. When it says "watch how the animation changes..." I'm expecting another DartPad. But, as you are probably thinking, do we NEED another DP here? I think not, but I'd rephrase this to say something like, "make the following change to the previous DartPad to see how the..."

Comment (legalcodes) [reply src/docs/codelabs/implicit-animations.md]: This was bothering me, you offered the perfect phrase for the job, many thanks Shams!

File: src/docs/codelabs/implicit-animations.md [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+## Example: AnimatedOpacity
+
+The following example demonstrates how to add a fade-in effect to existing UI
+using an implicitly animated widget called [AnimatedOpacity]. **The example begins
+with no animation code** – it consists of a [Material App] home screen containing:
+- A photograph of an owl.
+- One "Show Details" button that does nothing when clicked.
+- Description text of the owl in the photograph.
+
+Run the example to render the following code:
```
[the remark this answers is not in the document]
Comment (legalcodes) [reply src/docs/codelabs/implicit-animations.md]: Per our discussion, I've raised https://github.com/dart-lang/dart-pad/issues/1340

Could you add a screenshot and any additional description you might have of this issue?

Many thanks!

File: src/docs/codelabs/implicit-animations.md [hunk 8]
```diff
@@ -0,0 +1,472 @@
...
+  src="{{site.custom.dartpadx.embed-flutter-prefix}}?id=8501583cb789504d75317a5ba1ca6930"
+  style="border: 1px solid lightgrey; margin-top: 10px; margin-bottom: 25px"
+  frameborder="no" height="500" width="100%"
+></iframe>
+
+### Animate color, borderRadius, and margin with AnimatedContainer
+{{site.alert.secondary}}
+  This section contains a list of steps you can use to add an implicit animation to the
+  preceding example code. After the steps, you can also run an example with the
+  the changes already made.
```
Comment (goderbauer) [inline src/docs/codelabs/implicit-animations.md]: remove the "the"? It's already on the preceding line.

File: src/docs/codelabs/implicit-animations.md [hunk 9]
```diff
@@ -0,0 +1,472 @@
...
+{{site.alert.end}}
+
+## What are implicit animations?
+
+With Flutter's animation library, you can add motion and create visual effects
+for the widgets in your UI. One part of the library is an assortment of widgets
+that manage animations for you. These widgets are collectively referred to as _implicit
+animations_, or _implicitly animated widgets_, deriving their name from the
+[ImplicitlyAnimatedWidget] class that they implement. Implicit animations trade
+control for convenience&mdash;they manage animation effects so that you don't have to.
```
Comment (goderbauer) [inline src/docs/codelabs/implicit-animations.md]: I always like to say that with implicit animations you you set a target value and when that target value changes the widget animates you from the old to the new.

File: src/docs/codelabs/implicit-animations.md [hunk 10]
```diff
@@ -0,0 +1,472 @@
...
+- `AnimatedOpacity` listens for state changes in its `opacity` property.
+- Whenever `opacity` changes, `AnimatedOpacity` automatically animates the
+widget's transition to the new value for `opacity`.
+- `AnimatedOpacity` requires a `duration` parameter to define the time it takes
+to animate the transition between an old `opacity` value and a new one.
+
+The previous example also uses a common workflow for using implicit animations:
+- First, pick a widget property to animate.
+- Next, choose an implicit animation that can animate that property.
+- Set the start and end values for the property that you're animating.
```
Comment (goderbauer) [inline src/docs/codelabs/implicit-animations.md]: This sounds kinda odd. You're really only picking a start value. In the last bullet point, where you trigger the animation you basically trigger it by picking the end value.

File: examples/animation/implicit/container5/lib/main.dart [hunk 11]
```diff
@@ -0,0 +1,87 @@
+import 'package:flutter_web/material.dart';
+import 'package:flutter_web_test/flutter_web_test.dart';
+import 'package:flutter_web_ui/ui.dart' as ui;
+import 'dart:math';
+const _duration = Duration(milliseconds: 400);
```
Comment (goderbauer) [inline examples/animation/implicit/container5/lib/main.dart]: nit: add an empty line above this line.

File: examples/animation/implicit/container0/lib/main.dart [hunk 12]
```diff
@@ -0,0 +1,67 @@
+import 'package:flutter_web/material.dart';
```
Comment (goderbauer) [inline examples/animation/implicit/container0/lib/main.dart]: Do you need license headers for all the .dart files?

File: examples/animation/implicit/container0/lib/main.dart [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+import 'package:flutter_web_test/flutter_web_test.dart';
+import 'package:flutter_web_ui/ui.dart' as ui;
+import 'dart:math';
```
Comment (goderbauer) [inline examples/animation/implicit/container0/lib/main.dart]: There should be a blank line between package: and dart: imports. Also, the dart: imports should be before the package: imports.

File: examples/animation/implicit/container0/lib/main.dart [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+double randomMargin() {
+  return Random().nextDouble() * 64;
+}
+
+Color randomColor() {
+  return Color(0xFFFFFFFF & Random().nextInt(0xFFFFFFFF));
+}
+
+class AnimatedContainerDemo extends StatelessWidget {
+    final color = randomColor();
```
Comment (goderbauer) [inline examples/animation/implicit/container0/lib/main.dart]: nit: indent these lines by only two spaces.

File: examples/animation/implicit/container0/lib/main.dart [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    final borderRadius = randomBorderRadius();
+    final margin = randomMargin();
+
+  Widget build(BuildContext context) {
```
Comment (goderbauer) [inline examples/animation/implicit/container0/lib/main.dart]: Add: `@override`

File: examples/animation/implicit/container1/lib/main.dart [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+class AnimatedContainerDemo extends StatefulWidget {
+    _AnimatedContainerDemoState createState() => _AnimatedContainerDemoState();
+}
+
+class _AnimatedContainerDemoState extends State<AnimatedContainerDemo> {
+  Color color;
+  double borderRadius;
+  double margin;
+
+  initState() {
```
Comment (goderbauer) [inline examples/animation/implicit/container1/lib/main.dart]:
```
@override
void initState() {
  super.initState();

```

File: examples/animation/implicit/container1/lib/main.dart [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    final color = randomColor();
```
Comment (goderbauer) [inline examples/animation/implicit/container1/lib/main.dart]: remove the "final" on these lines? Not sure what that even means...

File: examples/animation/implicit/container7/lib/main.dart [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+          ),
+          onPressed: () => null),
+      Container(
+          child: Column(
+            children: <Widget>[
+              Text('Type: Owl'),
+              Text('Age: 39'),
+              Text('Employment: None'),
+            ],
+          ))
```
Comment (goderbauer) [inline examples/animation/implicit/container7/lib/main.dart]: nit:
```suggestion
          ),
        ),
```

File: examples/animation/implicit/container7/lib/main.dart [hunk 19]
```diff
@@ -0,0 +1,43 @@
+import 'package:flutter_web/material.dart';
+import 'package:flutter_web_test/flutter_web_test.dart';
+import 'package:flutter_web_ui/ui.dart' as ui;
+
+const owl_url =
```
Comment (goderbauer) [inline examples/animation/implicit/container7/lib/main.dart]: nit: I would just leave this on one line (unless you are bound to 80 chars per line)

File: examples/animation/implicit/opacity0/lib/main.dart [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+    ]);
+  }
+}
+
+class MyApp extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    return MaterialApp(
+        home: Scaffold(
+      body: Center(child: FadeInDemo()),
```
Comment (goderbauer) [inline examples/animation/implicit/opacity0/lib/main.dart]: Indentation is off here. This should probably be:

```dart
    return MaterialApp(
      home: Scaffold(
        body: Center(child: FadeInDemo()),
      ),
    );
```

File: src/docs/codelabs/implicit-animations.md (hunk 6 above)
Comment (goderbauer) [reply src/docs/codelabs/implicit-animations.md]: Lol, I also was looking for another dartpad here... But yeah, telling people to modify the previous one sounds good.

File: src/docs/codelabs/implicit-animations.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (goderbauer) [reply src/docs/codelabs/implicit-animations.md]: For me, dartpad was working fine.

File: src/docs/codelabs/implicit-animations.md (hunk 9 above)
Comment (legalcodes) [reply src/docs/codelabs/implicit-animations.md]: Perfect! This adds some much-needed glue the intro was missing. Many thanks.

File: src/docs/codelabs/implicit-animations.md (hunk 10 above)
Comment (legalcodes) [reply src/docs/codelabs/implicit-animations.md]: Yes, I think you're right -- `Set the start and end values` is misleading. The "workflow" I'm summarizing here is essentially a series of decisions made at design time, so `Choose the start and end values...` would be the better phrase. WDYT?

**Comment by galeyang** [comment]
**Like**
- Overall, it’s very clear on what the step is about, what code to change
- Like clear expectations: prerequirest, covered topics, est time
- Static code differences are very clear
- Like the animated curve graph

**Potential improvements**
- I kind of wish that I see this part earlier. I tried to summarize the high-level flow when I was reading the detailed steps. This part gives me a very clear overview. 
image

- I’m not sure if we should use [auto-run](https://github.com/dart-lang/dart-pad/pull/1317) in some of reading-centric examples. What value do users get from spending effort to click the run button? 
- Trivial: Would be helpful to remind me where to add the code. `The preceding examples do not specify a curve, so the implicit animations apply a linear animation curve by default. Add a curve parameter to the preceding example and watch how the animation changes when you pass the easeInOutBack constant for curve:`
- Nit: text said "Show details" but the code is “Show Details”

**Wishes** 
Some notes are outside of your scope but are general notes that we can improve in other PRs)
- I wish that I can copy code snippets in those code difference blocks more easily
   - When I wanted to practice these steps, I tried to copy the snippet to the DartPad above but it’s hard to do that. Issues (1) got irrelevant syntax (e.g. the plug sign) (2) Lots of scrolling. Eventually, I decided to read through the code
- (Trivial) Wish that the different code is highlighted in DartPad
- (Trivial) It would be good that we sometimes use Cupertino style (in new codelabs)
- (Trivial) For DartPad, I wish that there is a small hint to let me know the output is not recompiled based on my updated code yet. 

**General question**: what’s the difference between codelab and cookbook?

File: src/docs/codelabs/implicit-animations.md [hunk 21]
```diff
@@ -0,0 +1,472 @@
...
+
+The following example shows how to use the [AnimatedContainer] widget to
+animate multiple properties (`margin`, `borderRadius`, and `color`) with
+different types (`double` and `Color`).
+**The example begins with no animation code**&mdash;it starts with a [Material App]
+home screen that contains:
+
+- A shape with size, margin, and color that are different each time you run the
+example.
+- A **Change** button that does nothing when clicked.
```
[the remark this answers is not in the document]
Comment (legalcodes) [reply src/docs/codelabs/implicit-animations.md]: Sure -- how about something like:
```
Remember, an implicit animation is a subtype of `ImplicitlyAnimatedWidget`, and you can
browse the [list of common implicitly animated widgets] that ship with Flutter.
```

File: src/docs/codelabs/implicit-animations.md [hunk 22]
```diff
@@ -0,0 +1,472 @@
...
+The previous example demonstrates useful features of `AnimatedOpacity`:
+- `AnimatedOpacity` listens for state changes in its `opacity` property.
+- Whenever `opacity` changes, `AnimatedOpacity` automatically animates the
+widget's transition to the new value for `opacity`.
+- `AnimatedOpacity` requires a `duration` parameter to define the time it takes
+to animate the transition between an old `opacity` value and a new one.
+
+The previous example also uses a common workflow for using implicit animations:
+- First, pick a widget property to animate.
+- Next, choose an implicit animation that can animate that property.
```
[the remark this answers is not in the document]
Comment (legalcodes) [reply src/docs/codelabs/implicit-animations.md]: Sure -- how about something like:
```
Remember, an implicit animation is a subtype of `ImplicitlyAnimatedWidget`, and you can
browse the [list of common implicitly animated widgets] that ship with Flutter.
```

File: src/docs/codelabs/implicit-animations.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (legalcodes) [reply src/docs/codelabs/implicit-animations.md]: Let's followup to make sure I've fully captured this suggestion in the next change.

**Comment by legalcodes** [comment]
Awesome @galeyang, per our discussion I plan to make the following changes: 
* Distribute the "workflow" summary points throughout each step
* Wait for a later PR to use auto-run if we find it necessary
* Use named-examples with links, and also provide a hint to the user to use the provided line numbers for reference as the user makes suggested changes
* Fix the nit

**Comment by sfshaza2** [comment]
LGTM
